### DISCUSSION TITLE
CORS

### DISCUSSION THREAD

**Discussion by LeeThompson** [nicolargo/glances#2802] [Q&A] [answered]
... `has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

Is there any way to have Glances (running on Docker) to support CORS?

**Comment by nicolargo** [comment]
Hi @LeeThompson,

Following CORS rules are appplied in Glances:

```
        # FastAPI Enable CORS
        # https://fastapi.tiangolo.com/tutorial/cors/
        self._app.add_middleware(
            CORSMiddleware,
            # allow_origins=["*"],
            allow_origins=[self.bind_url],
            allow_credentials=True,
            allow_methods=["*"],
            allow_headers=["*"],
        )
```

Bind address can be set thanks to the --bind <mybindaddress> option (default is 0.0.0.0).

**Reply by LeeThompson** [reply]
Where's this set? I'm using the official docker image (`nicolargo/glances:latest`)
`GLANCES_OPT` is set to `-w`

**Reply by LeeThompson** [reply]
@nicolargo Is there a way to have the `allow_origins=["*"]` setting in the official Docker container?   That would probably fix this.   Binding is too limiting and does not solve the issue I'm having.    If you supported doing "`Allow-Origins: "*.domain"`" that could work as well although I think CORS locks in the *port* as well.

Currently, the only CORS Glances is returning from the jquery script is `access-control-allow-credentials: true` and nothing else.

**Reply by nicolargo** [reply]
@LeeThompson yes i can give a try on the develop branch (docker image: nicolargo/glances:dev). I will try to do it as soon as possible. I keep you inform. If it works a more advanced patch will be done to make this key configurable from the Glances configuration file.

**Reply by LeeThompson** [reply]
Thanks for taking a look!

**Reply by nicolargo** [reply]
Ça you try the docker develop tag  ?

**Reply by LeeThompson** [reply]
It is working!
No CORS errors being thrown.

```
HTTP/1.1 200 OK
date: Wed, 29 May 2024 17:13:38 GMT
server: uvicorn
content-length: 885
content-type: application/json
access-control-allow-origin: *
access-control-allow-credentials: true
```

**Reply by nicolargo** [reply, marked as answer]
Perfect.
My proposal is to set the default configuration to '*' and make it configurable through a New key in the Glances config file.

Any others thing needed to be configurable for the CORS ?

**Reply by LeeThompson** [reply]
Not that I know of, but I am not an expert although I've done an amazing amount of reading over the last few days lol

**Reply by LeeThompson** [reply]
(Marking this answered since a solution is coming to` :latest`)

**Comment by nicolargo** [comment]
-e GLANCES_OPT="-w --bind <IP@>"

**Reply by LeeThompson** [reply]
> -e GLANCES_OPT="-w --bind <IP@>"

The binding isn't the issue, I have _no_ idea where to put that `self._app.add_middleware(` block you posted.

Never even heard of FastAPI

**Comment by nicolargo** [comment]
The block is already in the Glances code, you do not have to include it in your configuration.

What is you use case ? I do not have this issue on my side.

PS: FastAPI is the framework used by Glances to expose the Rest API and the WebUI.

**Reply by LeeThompson** [reply]
My use case is a dashboard tool called Dasherr (https://github.com/erohtar/Dasherr) which has a widget that uses Glances.

Locally, I updated Dasherr to use the API v4 call (the author is aware of the issue and is working on it) and it should've fixed it - but my browsers (Chrome, Firefox, etc) were loaded with CORS errors.   

I added CORS settings to my Apache v2.4 config, no change.   

Right now I'm using Traefik as a workaround (it's setting the CORS stuff) but I'd like to do the fix on the Glances side.

**Reply by erohtar** [reply]
RE: https://github.com/nicolargo/glances/issues/479
Has this been changed/reversed in a recent update? Accessing Glances API using a simple JS request (no proxy) used to work under older Glances versions, but not anymore

**Comment by LeeThompson** [comment]
I tried modifying one of my Glances instances to use the `--bind`  parameter, I used the fully qualified domain name since the IP is the same as the host (it uses the 'host' docker network).  This is the same FQDN as Dasherr's API call, and I get the same CORS error.

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://redacted:61208/api/4/quicklook. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.`

**Reply by LeeThompson** [reply]
I will admit I am not that knowledgeable of CORS.   That out of that way, here is a comparison of Firefox response headers via Dasherr (the application listed above).  The first has it using Traefik's CORS settings, the second is direct from Glances.

These are both going to the same Glances instance using the `api/v4/quicklook` endpoint.

### Traefik 
```
HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Content-Encoding: gzip
Content-Length: 436
Content-Type: application/json
Date: Tue, 28 May 2024 08:14:09 GMT
Server: uvicorn
Vary: Accept-Encoding
```

### Glances (direct to :61208)
```
HTTP/1.1 200 OK
date: Tue, 28 May 2024 08:14:38 GMT
server: uvicorn
content-length: 441
content-type: application/json
access-control-allow-credentials: true
content-encoding: gzip
vary: Accept-Encoding
```

This to me (and I admit I am probably wrong about this) seems like Glances is not sending the CORS header elements at all (Access-Control-Allow-Credentials, Access-Control-Allow-Origin) for whatever reason.  It is sending the allow credentials although it's all lower case, I'm assuming that doesn't matter.

Dasherr is using jquery.js to actually send the initial request, it is setting Allow-Origin:

```
	request.setRequestHeader("X-Requested-With", "XMLHttpRequest");
	request.setRequestHeader("Access-Control-Allow-Origin", "*");
```
